[MRG+1] Kernel Ridge Regression

jmetzen opened the pull request:
Adds an implementation of Kernel Ridge Regression (KRR)

Basic implementation of KRR based on @mblondel's code in lightning. In addition to the lighning code, documentation and an example comparing KRR and SVR is added. The example illustrates the models learned by KRR and SVR:

![plot_kernel_ridge_regression_001](https://cloud.githubusercontent.com/assets/1116263/5330658/9793f8f6-7e03-11e4-93a9-7da2f0634037.png)

Furthermore, it compares the time required for fitting the model and predicting with the model for different sizes of the training set:

![plot_kernel_ridge_regression_002](https://cloud.githubusercontent.com/assets/1116263/5330660/a1b7a4cc-7e03-11e4-8628-3d6cccad921a.png)

Commit:
ADD @mblondel's initial implementation of KernelRidge added

Commit:
ADD Example comparing kernel ridge and support vector regression

Commit:
TEST Added tests for KernelRidge based on @mblondel's code in lightning

Commit:
DOC Added documentation for kernel_ridge module

Commit:
ENH Example comparing KRR and SVR documented and slightly modified

Commit:
FIX Fixed assigning colors to methods in KRR example

Commit:
DOC Added narrative doc for kernelized ridge regreesion

Commit:
DOC Resolved minor issues in documentation of kernel_ridge

Commit:
MISC Optimized two numpy statements in kernel_ridge.py

Commit:
FIX Add kernel_ridge to list of all submodules

Commit:
FIX Input validation in fit() of KernelRidge

Commit:
FIX Allow multi-output in KernelRidge

Commit:
REFACTOR Backported changes in _solve_cholesky_kernel to ridge.py

mblondel commented on doc/modules/kernel_ridge.rst:
```diff
@@ -0,0 +1,55 @@
+function in the original space.
+
+The model learned by :class:`KernelRidge` is identical to support vector regression
+(:class:`SVR`). However, different loss functions are used (ridge versus
```
squared loss is a more common name

jmetzen replied:
in my understanding, squared loss doesn't include L2 regularization. I will rephrase it as "ridge
loss, i.e., squared error loss combined with L2 regularization, ..."; that should clarify the point.

GaelVaroquaux replied:
> in my understanding, squared loss doesn't include L2 regularization. I will
> rephrase it as "ridge
> loss, 

A loss never includes the regularization term. It's the defination of
loss.

jmetzen replied:
you are right; the term ridge loss used in some presentations on ridge regression  (e.g., http://statweb.stanford.edu/~nzhang/203_web/lecture12_2010.pdf, slide 6) doesn't make sense. I will change that part

eickenberg replied:
least squares loss with ridge penalty / l2 regularization term

mblondel replied:
We call it squared loss in the SGD module.

mblondel commented on doc/modules/kernel_ridge.rst:
```diff
@@ -0,0 +1,55 @@
+non-kernelized Ridge in situations when the number of feature dimensions D is
+considerably larger than the number of training datapoints N since the
+computational cost of solving the dual (kernelized) problem is :math:`O(N^3)`
+while the standard Ridge requires :math:`O(D^3)`.
```
Indeed you can remove this paragraph since this is already taken care of in `Ridge`.

mblondel commented on doc/modules/kernel_ridge.rst:
```diff
@@ -0,0 +1,55 @@
+epsilon-insensitive loss). In contrast to :class:`SVR`, fitting
+:class:`KernelRidge` can be done in closed-form and is typically faster for
+medium-sized datasets. On the other hand, the learned model is non-sparse and
+thus slower than SVR at prediction-time.
```
Provided that epsilon > 0. The solution is dense when epsilon=0.

mblondel commented on doc/modules/kernel_ridge.rst:
```diff
@@ -0,0 +1,55 @@
+faster than :class:`KernelRidge` for all sizes of the training set because of
+the learned sparse solution. Note that the degree of sparsity and thus the
+prediction time depends on the parameters :math:`epsilon` and :math:`C` of the
+:class:`SVR`.
```
Perhaps you can include my remark about epsilon here.

mblondel commented on sklearn/linear_model/ridge.py:
```diff
@@ -141,7 +144,14 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
 
-        dual_coef = linalg.solve(K, y, sym_pos=True, overwrite_a=True)
+        try:
+            # XXX: Check if overwrite_a=True is ok when copy is False
```
Can you address this?

jmetzen replied:
I think, we have to change that to overwrite_a=False. In case that a LinAlgError is raised and the pseudo-inverse of K is computed instead, the pseudo-inverse would be based on the modified K.

mblondel commented:
+1 for merge on my side once my (minor) comments are addressed. Would be nice to have a second review and merge this soon. I am planning to remove it from lightning once this is merged.

eickenberg commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -0,0 +1,172 @@
+from sklearn.kernel_ridge import KernelRidge
+import matplotlib.pyplot as plt
+
+np.random.seed(0)
```
Generally the use of `rng = np.random.RandomState(seed)` is preferred. Random numbers are then generated with e.g. `gauss = rng.randn(shape)`

eickenberg commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -0,0 +1,172 @@
+
+#############################################################################
+# Generate sample data
+X = 5 * np.random.rand(10000, 1)
```
`rng.rand(10000, 1)`

eickenberg commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -0,0 +1,172 @@
+y = np.sin(X).ravel()
+
+# Add noise to targets
+y[::5] += 3 * (0.5 - np.random.rand(X.shape[0]/5))
```
`rng.rand`

eickenberg commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -0,0 +1,172 @@
+plt.figure()
+
+# Generate sample data
+X = 5 * np.random.rand(10000, 1)
```
`rng.rand`

eickenberg commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -0,0 +1,172 @@
+# Generate sample data
+X = 5 * np.random.rand(10000, 1)
+y = np.sin(X).ravel()
+y[::5] += 3 * (0.5 - np.random.rand(X.shape[0]/5))
```
`rng.rand`

eickenberg commented on sklearn/kernel_ridge.py:
```diff
@@ -0,0 +1,185 @@
+    function in the original space.
+
+    The model learned by KRR is identical to support vector regression (SVR).
+    However, different loss functions are used (ridge versus  epsilon-
```
least squares versus epsilon-insensitive

eickenberg commented on sklearn/kernel_ridge.py:
```diff
@@ -0,0 +1,185 @@
+    the learned model is non-sparse and thus slower than SVR at prediction-time.
+
+    Using KRR with a linear kernel can be advisable over non-kernelized Ridge
+    in situations when the number of feature dimensions D is considerably larger
```
Nitpick: I think this is preferably represented by `P` instead of `D`. Same thing three lines lower

eickenberg commented on sklearn/kernel_ridge.py:
```diff
@@ -0,0 +1,185 @@
+    >>> from sklearn.kernel_ridge import KernelRidge
+    >>> import numpy as np
+    >>> n_samples, n_features = 10, 5
+    >>> np.random.seed(0)
```
`rng = np.random.RandomState(0)` etc

eickenberg commented on sklearn/linear_model/ridge.py:
```diff
@@ -123,14 +123,17 @@ def _solve_cholesky(X, y, alpha, sample_weight=None):
 
 
-def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
+def _solve_cholesky_kernel(K, y, alpha, sample_weight=1.0, copy=False):
```
I think we had come to the conclusion that it was better to have `sample_weight=None` as default when there are no sample weights. I'll try to find the discussion.

Commit:
MISC Using RandomState object instead of setting global seed

Commit:
DOC Revised documentation of kernel ridge regression

eickenberg commented on sklearn/linear_model/ridge.py:
```diff
@@ -141,7 +144,14 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
+        except np.linalg.LinAlgError:
+            warnings.warn("Singular matrix in solving dual problem. Using "
+                          "pseudo inverse instead.")
+            Kinv = linalg.pinv2(K)
```
I am sorry, I am unable to find the comments in the diff when this was still in the kernel ridge function. So my comment may have been handled, but I'd prefer to make sure: How small does `alpha` have to be and for what examples of `K` does this situation crop up? Wouldn't using a least squares solver (`scipy.linalg.lstsq`) in analogy to the `try` block be a semantically better choice? (again, really sorry if this has already been addressed/dismissed)

amueller replied:
Also wondering this....

amueller commented:
I'm not super excited about making this top-level, but I also don't have a better idea :-/
It could go into linear model....

amueller commented on doc/modules/kernel_ridge.rst:
```diff
@@ -0,0 +1,57 @@
+
+.. currentmodule:: sklearn.kernel_ridge
+
+Kernelized ridge regression (KRR) [M2012]_ combines ridge regression (linear 
```
maybe link to the ridge regression docs?

amueller commented on doc/modules/kernel_ridge.rst:
```diff
@@ -0,0 +1,57 @@
+the data. For non-linear kernels, this corresponds to a non-linear
+function in the original space.
+
+The model learned by :class:`KernelRidge` is identical to support vector
```
Maybe say "the form of the model" or "the parametrization of the model"? I feel like "the model learned" also includes the learned parameters, which will be different.

amueller commented on doc/modules/kernel_ridge.rst:
```diff
@@ -0,0 +1,57 @@
+The model learned by :class:`KernelRidge` is identical to support vector
+regression (:class:`SVR`). However, different loss functions are used: KRR uses
+squared error loss combined with l2 regularization while support vector
+regression uses :math:`\epsilon`-insensitive loss.  In contrast to :class:`SVR`,
```
Both use l2 regularization, right?

amueller commented on doc/modules/kernel_ridge.rst:
```diff
@@ -0,0 +1,57 @@
+   :target: ../auto_examples/plot_kernel_ridge_regression.html
+   :align: center
+
+The next figure compares the time for fitting and prediction of
```
Very nice explanations :)

amueller commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -0,0 +1,172 @@
+# Fit regression model
+train_size = 100
+svr = GridSearchCV(SVR(kernel='rbf', gamma=1e-1), cv=5,
+				   param_grid={"C": [1e0, 1e1, 1e2, 1e3],
```
maybe fix the indentation here?

eickenberg replied:
and the potential PEP8 violation in the subsequent lines (> 80 columns?)

amueller commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -0,0 +1,172 @@
+y[::5] += 3 * (0.5 - rng.rand(X.shape[0]/5))
+sizes = np.logspace(1, 4, 7)
+for name, estimator in {"KRR": KernelRidge(kernel='rbf', alpha=1e-1, 
+	                                      gamma=1e+1),
```
Interesting way to write 10 ;)

amueller commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -0,0 +1,172 @@
+	train_time = []
+	test_time = []
+	for train_test_size in sizes:
+		t0 = time.time()
```
We should really return that from the learning curves.....

amueller commented on sklearn/kernel_ridge.py:
```diff
@@ -0,0 +1,181 @@
+    def _pairwise(self):
+        return self.kernel == "precomputed"
+
+    def fit(self, X, y=None, sample_weight=1.0):
```
Should sample_weight default to ``None``?

amueller commented on sklearn/tests/test_kernel_ridge.py:
```diff
@@ -0,0 +1,42 @@
+def test_kernel_ridge():
+    pred = Ridge(alpha=1, fit_intercept=False).fit(X, y).predict(X)
+    pred2 = KernelRidge(kernel="linear", alpha=1).fit(X, y).predict(X)
+    assert_array_almost_equal(pred, pred2)
```
could you maybe use decision function instead?

jmetzen replied:
Turns out KernelRidge does not implement decision_function yet. Before adding it: what's actually the difference between predict and decision_function in the regression case?

amueller replied:
Oh right, I'm sorry. For regression they are the same. You don't need to add it. I think we want to add it to all regressors in the future but that might be a different issue.

amueller commented on sklearn/tests/test_kernel_ridge.py:
```diff
@@ -0,0 +1,42 @@
+
+def test_kernel_ridge_precomputed():
+    K = np.dot(X, X.T)
+    pred = KernelRidge(kernel="linear").fit(X, y).predict(X)
```
I'd also prefer decision function here. predict is a very rough approximation of "the same mode".

amueller commented on sklearn/tests/test_kernel_ridge.py:
```diff
@@ -0,0 +1,42 @@
+    K = np.dot(X, X.T)
+    K2 = K.copy()
+    sw = np.ones(X.shape[0]) / float(X.shape[0])
+    KernelRidge(kernel="precomputed").fit(K, y, sample_weight=sw)
```
That seems more like a smoke test... can you maybe also test against a linear Ridge?

jmetzen replied:
this test originates from lightning. @mblondel: the only thing it seems to check is that K is not changed per side-effect during fit, right? In that case I would propose to rename the test to test_kernel_ridge_precomputed_kernel_unchanged.

amueller replied:
makes sense. checking the sample weights in the precomputed case would also be nice but I don't want to press it.

mblondel replied:
Indeed, the test could be strengthened by comparing with Ridge (although
this could be kind of circular since they use the same underlying function
in the n_samples < n_features case).

Commit:
DOC Polishing doc of kernel_ridge

Commit:
PEP8 Removing PEP8 violations in plot_kernel_ridge_regression.py

Commit:
MISC Let sample_weight default to None

Commit:
FIX We must not use overwrite_a=True in _solve_cholesky_kernel as K might be reused

jmetzen commented on sklearn/linear_model/ridge.py:
```diff
@@ -141,7 +145,17 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
+            warnings.warn("Singular matrix in solving dual problem. Using "
+                          "pseudo inverse instead.")
+            Kinv = linalg.pinv2(K)
+            dual_coef = np.dot(Kinv, y)
```
Re-adding @eickenberg's comment that got once again outdated:
I am sorry, I am unable to find the comments in the diff when this was still in the kernel ridge function. So my comment may have been handled, but I'd prefer to make sure: How small does alpha have to be and for what examples of K does this situation crop up? Wouldn't using a least squares solver (scipy.linalg.lstsq) in analogy to the try block be a semantically better choice? (again, really sorry if this has already been addressed/dismissed)

jmetzen replied:
From the history, @eickenberg also commented the following:
np.linalg.lstsq accepts multiple targets. However, afaik it does an SVD. So that would be almost as costly as the full pseudoinverse. I am starting to understand :) But I'd still prefer a least squares solver.

jmetzen replied:
what about allowing to specify different solvers as in ridge?

amueller replied:
Why not. What would ``auto`` be then? I'm surprised that least squares is never selected in ridge by auto.

eickenberg replied:
wrt my comment: calling `pinv2` or `lstsq` basically boils down to doing an svd on `K + ridge`. So from that point of view it should be the one or the other. However, looking at the underlying lapack functions used, `lstsq` calls `gelss` ("solve least squares using SVD") and `pinv2` calls scipy svd which in turn calls `gesdd` ("do svd using a divide and conquer method"). What I can't discern is whether maybe `lstsq` manages to avoid some calculation, and also on what svd implementation it is based. In any case: I don't think it makes sense to give the option to decide between `pinv2` and `lstsq` and in doubt it should be the function dedicated to solving this type of problem, i.e. `lstsq`.

As far as other solvers goes, yes, there are other ways of solving kernel ridge, e.g. by re-using an eigendecomposition of `K` for potentially different penalties to name only one. But shouldn't adding more solvers and decision logic should maybe be kept for after merging this? (I think this code already provides enough functionality to be considered a valid contribution.)

Commit:
TEST Extending tests for kernel_ridge
 
* Testing more precomputed kernels
* Testing sample_weights and comparing to Ridge

Commit:
REFACTOR Using lstsq rather than manual pinv in _solve_cholesky_kernel

Commit:
MINOR Fixed typo in test name (test_kernel_ridge_singular_kernel)

jmetzen commented:
I've added an additional test "test_kernel_ridge_sample_weights", which compares Ridge, KernelRidge and KernelRidge with precomputed kernel when sample-weights are given.

Regarding the solvers: I agree, we can leave alternative solvers to future PRs. I changed the fallback in _solve_cholesky_kernel to linalg.lstsq as @eickenberg suggested. The additional test "test_kernel_ridge_singular_kernel" compares the solution of this fallback to Ridge with solver="lsqr". The results are equal to Ridge's solution only in the first two decimals, though.  Ridge gives also different results depending on the the chosen solver.

Commit:
TEST Use make_regression instead of make_classification when testing kernel_ridge

Commit:
DOC Added KernelRidge to whats_new.rst

Commit:
FIX KernelRidge checks if fitted before predict

jmetzen commented:
I've rebased to support check_is_fitted in KernelRidge. From my side, all todos are completed now and we could merge.

@mblondel I've added KernelRidge to whats_new.rst and listed you as (co-)author since this PR is heavily based on your original code from lightning.

amueller commented on sklearn/linear_model/ridge.py:
```diff
@@ -141,7 +145,16 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
+        except np.linalg.LinAlgError:
+            warnings.warn("Singular matrix in solving dual problem. Using "
+                          "least-squares solution instead.")
+            dual_coef = linalg.lstsq(K, y)[0]
```
What is the reason to not use ``lstsq`` for all cases? Is it slower or does it provide a lower quality solution? I still didn't get that from your comment.

eickenberg replied:
the normal equations should have an exact solution. the fact that
sympos=True means that it uses the optimized posv. however, least squares
gives the same solution, of course. i don't know which is faster, although
the latter calculates an svd which the former probably doesn't.

On Tuesday, January 13, 2015, Andreas Mueller <notifications@github.com>
wrote:

> In sklearn/linear_model/ridge.py
> <https://github.com/scikit-learn/scikit-learn/pull/3942#discussion-diff-22891602>
> :
>
> > @@ -141,7 +145,16 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
> >          # Only one penalty, we can solve multi-target problems in one time.
> >          K.flat[::n_samples + 1] += alpha[0]
> >
> > -        dual_coef = linalg.solve(K, y, sym_pos=True, overwrite_a=True)
> > +        try:
> > +            # Note: we must use overwrite_a=False in order to be able to
> > +            #       use the fall-back solution below in case a LinAlgError
> > +            #       is raised
> > +            dual_coef = linalg.solve(K, y, sym_pos=True,
> > +                                     overwrite_a=False)
> > +        except np.linalg.LinAlgError:
> > +            warnings.warn("Singular matrix in solving dual problem. Using "
> > +                          "least-squares solution instead.")
> > +            dual_coef = linalg.lstsq(K, y)[0]
>
> What is the reason to not use lstsq for all cases? Is it slower or does
> it provide a lower quality solution? I still didn't get that from your
> comment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3942/files#r22891602>.
>

mblondel replied:
One advantage is that solve does the Cholesky decomposition only once in the multi-output case. Which reminds me BTW that the except block is probably broken in the multi-output case.

amueller replied:
Hum that seems hard to test for. Or can we deterministically get to the try with an ill-conditioned problem? Exemplifies the kind of trouble you get into with these try hacks ;)

mblondel replied:
If I remember correctly, this test triggers the exception:
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/tests/test_ridge.py#L98

mblondel replied:
I actually don't like this except either. I think I would rather let the estimator crash and let the user change the solver manually.

mblondel replied:
Of course, this can be address in another PR, as this warrants further discussion.

jmetzen replied:
This test also triggers the LinAlgError in _solve_cholesky_kernel:
https://github.com/jmetzen/scikit-learn/blob/kernel_ridge/sklearn/tests/test_kernel_ridge.py#L21

jmetzen replied:
Out of curiosity, I did a comparison ot the execution time of "scipy.linalg.solve(K, y, sym_pos=True, overwrite_a=False)" and "scipy.linalg.lstsq(K, y)[0]" for different sizes of K:
![figure_1](https://cloud.githubusercontent.com/assets/1116263/5744184/a78f32b4-9c1e-11e4-9dcc-3151d3d9a332.png)
solve is thus between one and two orders of magnitude faster.

eickenberg replied:
beautiful! thanks for the effort!!

On Wednesday, January 14, 2015, Jan Hendrik Metzen <notifications@github.com>
wrote:

> In sklearn/linear_model/ridge.py
> <https://github.com/scikit-learn/scikit-learn/pull/3942#discussion-diff-22955171>
> :
>
> > @@ -141,7 +145,16 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
> >          # Only one penalty, we can solve multi-target problems in one time.
> >          K.flat[::n_samples + 1] += alpha[0]
> >
> > -        dual_coef = linalg.solve(K, y, sym_pos=True, overwrite_a=True)
> > +        try:
> > +            # Note: we must use overwrite_a=False in order to be able to
> > +            #       use the fall-back solution below in case a LinAlgError
> > +            #       is raised
> > +            dual_coef = linalg.solve(K, y, sym_pos=True,
> > +                                     overwrite_a=False)
> > +        except np.linalg.LinAlgError:
> > +            warnings.warn("Singular matrix in solving dual problem. Using "
> > +                          "least-squares solution instead.")
> > +            dual_coef = linalg.lstsq(K, y)[0]
>
> Out of curiosity, I did a comparison ot the execution time of
> "scipy.linalg.solve(K, y, sym_pos=True, overwrite_a=False)" and
> "scipy.linalg.lstsq(K, y)[0]" for different sizes of K:
> [image: figure_1]
> <https://cloud.githubusercontent.com/assets/1116263/5744184/a78f32b4-9c1e-11e4-9dcc-3151d3d9a332.png>
> solve is thus between one and two orders of magnitude faster.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3942/files#r22955171>.
>

jmetzen replied:
@mblondel: the except block is not broken in the multi-output case since lstsq also supports multi-output: http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.linalg.lstsq.html

amueller replied:
Just to make sure: since which scipy?

amueller replied:
@jmetzen thanks for the graph, awesome :)

jmetzen replied:
at least since scipy version 0.9: http://docs.scipy.org/doc/scipy-0.9.0/reference/generated/scipy.linalg.lstsq.html

eickenberg replied:
@jmetzen if and only if it is easy to add to your bench: For completeness, the runtime of the `pinv2` option is interesting. Who knows, there may be a surprise :) 

Thanks also for checking back in scipy release history for the multitarget functionality of `lstsq`. It is true that I only checked on my computer with 0.14 before I proposed it.

jmetzen replied:
Sure, added a line for ""np.dot(scipy.linalg.pinv2(K), y)":
![figure_1](https://cloud.githubusercontent.com/assets/1116263/5746735/f9e93bbe-9c31-11e4-9d0c-91866482fc2f.png)
The differences between pinv and lstsq are probably not significant as the plot is based on a single repetition.

eickenberg replied:
Thanks again - reassuring that two methods based on SVD do not diverge completely in the constant hidden by O notation. To me this means we should be good to go.

mblondel replied:
@jmetzen Thanks. Maybe the documentation I saw was outdated then.

mblondel replied:
Nope, I was just confusing lsqr and lstsq.

amueller commented:
:+1: from my side.

eickenberg commented:
LGTM, too

jmetzen commented:
can we merge?

mblondel commented:
Merged by rebase. Thanks a lot @jmetzen!

If you feel motivated, you can also add KernelRidgeCV :). Again, the code is already there. It's just a matter of exposing it with a proper class.

jmetzen commented:
KernelRidgeCV would be interesting; I will take a look at it in the next days :-)

jmetzen commented:
PR #4118 fixes an issue of this PR (I forgot adding the new kernel_ridge module to classes.rst)

eickenberg commented:
Thanks @jmetzen for the coding effort and patience and @mblondel @amueller @GaelVaroquaux for the reviews!

As for `KernelRidgeCV`, I think it is a good idea to have it. As a matter of fact, whenever `Ridge(CV)` decides to use the kernelized version, it should probably start using these implementations, right?

Also, in order to put some pressure on myself, let it be stated that by the end of the coming weekend I intend to write something hopefully interesting about KernelRidgeCV.

jmetzen commented:
@eickenberg sounds great! In order to avoid that we both work on the same problem, I won't do too much until the end of the coming weekend ;-)

dougalsutherland commented:
@eickenberg @jmetzen Did you ever make any progress on KernelRidgeCV? I was just looking for it and got sad that it wasn't there.

eickenberg commented:
Do you mean analytical leave-one-out or something else?

I have a PR to fix RidgeCV that isn't merged ;)

No progress AFAIK, but should make time to do so.

On Wed, Feb 3, 2016 at 5:50 PM, Dougal J. Sutherland <
notifications@github.com> wrote:

> @eickenberg <https://github.com/eickenberg> @jmetzen
> <https://github.com/jmetzen> Did you ever make any progress on
> KernelRidgeCV? I was just looking for it and got sad that it wasn't there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3942#issuecomment-179340913>
> .
>

mblondel commented:
@dougalsutherland The functionality is essentially there since RidgeCV supports the n_features > n_samples case. It's just a matter of extracting the code and exposing it in an estimator class. PR welcome :)

jmetzen commented:
Would be nice to have this functionality but I don't have time at the moment to work on this. If someone comes up with a PR, I would be happy to participate in reviewing.
